File: Crypto/Android/cn/vimfung/luascriptcore/modules/crypto/Crypto.java
package cn.vimfung.luascriptcore.modules.crypto;

import android.os.Message;
import android.util.Log;

import java.io.UnsupportedEncodingException;
import java.security.DigestException;
import java.security.InvalidKeyException;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

import cn.vimfung.luascriptcore.LuaExportType;
import cn.vimfung.luascriptcore.modules.foundation.Encoding;

/**
 * 加密类
 */
public class Crypto implements LuaExportType
{
    /**
     * 对数据使用MD5
     * @param data 需要MD5的数据
     * @return MD5后的数据
     */
    public static byte[] md5(Object data)
    {
        if (data instanceof byte[] || data instanceof String)
        {
            byte[] rawData = null;
            if (data instanceof String)
            {
                rawData = ((String) data).getBytes();
            }
            else
            {
                rawData = (byte[]) data;
            }

            try
            {
                MessageDigest md = MessageDigest.getInstance("MD5");
                md.update(rawData);
                return md.digest();
            }
            catch (NoSuchAlgorithmException e)
            {
                e.printStackTrace();
            }

        }

        return null;
    }

    /**
     * 对数据使用SHA1
     * @param data 需要SHA1的数据
     * @return SHA1后的数据
     */
    public static byte[] sha1(Object data)
    {
        if (data instanceof byte[] || data instanceof String)
        {
            byte[] rawData = null;
            if (data instanceof String)
            {
                rawData = ((String) data).getBytes();
            }
            else
            {
                rawData = (byte[]) data;
            }

            try
            {
                MessageDigest md = MessageDigest.getInstance("SHA");
                md.update(rawData);
                return md.digest();
            }
            catch (NoSuchAlgorithmException e)
            {
                e.printStackTrace();
            }

        }

        return null;
    }

    /**
     * 对数据使用HMAC-MD5
     *
     * @param data 需要HMAC-MD5的数据
     * @param key 密钥
     * @return HMAC-MD5后的数据
     */
    public static byte[] hmacMD5(Object data, Object key)
    {
        byte[] rawKeyData = null;
        byte[] rawData = null;

        if (key instanceof byte[] || key instanceof String)
        {
            if (key instanceof String)
            {
                rawKeyData = ((String) key).getBytes();
            }
            else
            {
                rawKeyData = (byte[]) key;
            }
        }

        if (data instanceof byte[] || data instanceof String)
        {
            if (data instanceof String)
            {
                rawData = ((String) data).getBytes();
            }
            else
            {
                rawData = (byte[]) data;
            }
        }

        if (rawKeyData != null && rawData != null)
        {
            try
            {
                Mac mac = Mac.getInstance("HmacMD5");
                SecretKeySpec secret = new SecretKeySpec(
                        rawKeyData, "HmacMD5");
                mac.init(secret);
                byte[] result = mac.doFinal(rawData);

                return result;
            }
            catch (InvalidKeyException e)
            {
                e.printStackTrace();
            }
            catch (NoSuchAlgorithmException e)
            {
                e.printStackTrace();
            }
        }

        return null;
    }

    /**
     * 对数据使用HMAC-SHA1
     * @param key 密钥
     * @param data 需要HMAC-SHA1的数据
     * @return HMAC-SHA1后的数据
     */
    public static byte[] hmacSHA1(Object data, Object key)
    {
        byte[] rawKeyData = null;
        byte[] rawData = null;

        if (key instanceof byte[] || key instanceof String)
        {
            if (key instanceof String)
            {
                rawKeyData = ((String) key).getBytes();
            }
            else
            {
                rawKeyData = (byte[]) key;
            }
        }

        if (data instanceof byte[] || data instanceof String)
        {
            if (data instanceof String)
            {
                rawData = ((String) data).getBytes();
            }
            else
            {
                rawData = (byte[]) data;
            }
        }

        if (rawKeyData != null && rawData != null)
        {
            try
            {
                Mac mac = Mac.getInstance("HmacSHA1");
                SecretKeySpec secret = new SecretKeySpec(
                        rawKeyData, "HmacSHA1");
                mac.init(secret);
                byte[] result = mac.doFinal(rawData);

                return result;
            }
            catch (InvalidKeyException e)
            {
                e.printStackTrace();
            }
            catch (NoSuchAlgorithmException e)
            {
                e.printStackTrace();
            }
        }

        return null;
    }
}

File: Crypto/Unity3D/Crypto.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using System.Security.Cryptography;

namespace cn.vimfung.luascriptcore.modules.crypto
{

	/// <summary>
	/// 加密模块
	/// </summary>
	public class Crypto : object, LuaExportType 
	{
		public static byte[] md5(object data)
		{
			if (data is byte[] || data is string)
			{
				byte[] rawData = null;
				if (data is string)
				{
					rawData = System.Text.Encoding.UTF8.GetBytes (data as string);
				}
				else
				{
					rawData = data as byte[];
				}

				MD5CryptoServiceProvider md5Hasher = new MD5CryptoServiceProvider();
				return md5Hasher.ComputeHash(rawData);

			}

			return null;
		}

		public static byte[] sha1(object data)
		{
			if (data is byte[] || data is string)
			{
				byte[] rawData = null;
				if (data is string)
				{
					rawData = System.Text.Encoding.UTF8.GetBytes (data as string);
				}
				else
				{
					rawData = data as byte[];
				}
					
				SHA1CryptoServiceProvider shaHasher = new SHA1CryptoServiceProvider();
				return shaHasher.ComputeHash(rawData);

			}

			return null;
		}

		public static byte[] hmacMD5(object data, object key)
		{
			byte[] rawKey = null;
			byte[] rawData = null;

			if (data is byte[] || data is string)
			{
				if (data is string)
				{
					rawData = System.Text.Encoding.UTF8.GetBytes (data as string);
				}
				else
				{
					rawData = data as byte[];
				}
			}

			if (key is byte[] || key is string)
			{
				if (key is string)
				{
					rawKey = System.Text.Encoding.UTF8.GetBytes (key as string);
				}
				else
				{
					rawKey = key as byte[];
				}
			}

			if (data != null && key != null)
			{
				HMACMD5 hmac = new HMACMD5 (rawKey);
				return hmac.ComputeHash (rawData);
			}

			return null;
		}

		public static byte[] hmacSHA1(object data, object key)
		{
			byte[] rawKey = null;
			byte[] rawData = null;

			if (data is byte[] || data is string)
			{
				if (data is string)
				{
					rawData = System.Text.Encoding.UTF8.GetBytes (data as string);
				}
				else
				{
					rawData = data as byte[];
				}
			}

			if (key is byte[] || key is string)
			{
				if (key is string)
				{
					rawKey = System.Text.Encoding.UTF8.GetBytes (key as string);
				}
				else
				{
					rawKey = key as byte[];
				}
			}

			if (data != null && key != null)
			{
				HMACSHA1 hmac = new HMACSHA1 (rawKey);
				return hmac.ComputeHash (rawData);
			}

			return null;
		}
	}
}